Blog article for CloudWatch Metric Alarms (#30)
localstack/localstack.github.io: content/blog/2022-05-04-cloudwatch-metric-alarms/index.md

---
title: CloudWatch in Action - Get notified when your Lambda suddenly fails
description: "LocalStack now supports CloudWatch metric alarms! In this article, you will learn how to configure and test a simple AWS CloudWatch metric alarm with LocalStack, to get notified on infrastructure failures."
lead: "LocalStack now supports CloudWatch metric alarms! In this article, you will learn how to configure and test a simple AWS CloudWatch metric alarm with LocalStack, to get notified on infrastructure failures."
date: 2022-05-04
lastmod: 2022-05-04
draft: false
images: []

> contributors: ["Stefanie"]

Author: I think most articles were published as "LocalStack Team"?

Reviewer: i think the technical articles should highlight the main author!

tags: ["tutorial"]
---

> [AWS CloudWatch](https://docs.aws.amazon.com/cloudwatch/index.html) is a service that enables monitoring of your AWS infrastructure by collecting logs and operational metrics of your deployment. An integral part of infrastructure operations is reacting in real-time to anomalous changes in metrics, e.g., spiking disk or CPU usage, or failures in your serverless functions. This is where CloudWatch metric alarms and actions come into play, which you can now develop and test using LocalStack!
> 
> LocalStack now supports CloudWatch metric alarms with `statistic` and `comparison-operator`.
> In this article, we will discuss how you can use a CloudWatch metric alarm to get notified automatically when your Lambda function invocations fail. In our example, we will set up an email notification using the Simple Email Service (SES).

Reviewer 2: ```suggestion
[AWS CloudWatch](https://docs.aws.amazon.com/cloudwatch/index.html) is a service that enables monitoring of your AWS infrastructure by collecting logs and operational metrics of your deployment. An integral part of infrastructure operations is reacting in real-time to anomalous changes in metrics, e.g., spiking disk or CPU usage or failures in your serverless functions. This is where CloudWatch metric alarms and actions come into play, which you can now develop and test using LocalStack!

LocalStack now supports CloudWatch metric alarms with `statistic` and `comparison-operator`. This article will discuss how you can use a CloudWatch metric alarm to get notified automatically when your Lambda function invocations fail. We will set up an email notification using the Simple Email Service (SES) in our example.
```

## Prerequisites

For this tutorial you will need:
* [LocalStack Pro](https://localstack.cloud), to send emails via SMTP and SES.
* The [awslocal](https://docs.localstack.cloud/integrations/aws-cli/#localstack-aws-cli-awslocal) command line utility
* A mock SMTP server like [smtp4dev](https://github.com/rnwood/smtp4dev) or [Papercut SMTP](https://github.com/ChangemakerStudios/Papercut-SMTP) to receive the email notifications locally.

> In order to connect LocalStack with the SMTP server, you need to [configure the following SMTP environment variables](https://docs.localstack.cloud/aws/ses/#pro) when starting LocalStack:

Reviewer 2: ```suggestion
To connect LocalStack with the SMTP server, you need to [configure the following SMTP environment variables](https://docs.localstack.cloud/aws/ses/#pro) when starting LocalStack:
```

 * `SMTP_HOST` this should contain the hostname and the port of your mock SMTP server
 * `SMTP_USER` optional, if there is user to connect
 * `SMTP_PASS` optional

For example, when using smtp4dev, simply run:

    docker run --rm -it -p 3000:80 -p 2525:25 rnwood/smtp4dev

and set `SMTP_HOST=localhost:2525`.
Navigating to `http://localhost:3000` will open a UI to access the email notifications.

> ## CloudWatch Basics
> CloudWatch can help you to get a better understanding about how your AWS infrastructure resources behave over time. While some metrics are collected automatically, e.g., the execution of Lambdas, you can also define custom metrics you would like to monitor. To that end, you can use the CloudWatch API [`put-metric-data`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudwatch/put-metric-data.html).

Reviewer 2: ```suggestion
## CloudWatch Basics

CloudWatch can help you to get a better understanding of how your AWS infrastructure resources behave over time. While some metrics are collected automatically, e.g., the execution of Lambdas, you can also define custom metrics you would like to monitor. To that end, you can use the CloudWatch API [`put-metric-data`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudwatch/put-metric-data.html).
```

In order to create an alarm, you need to define what kind of metrics the alarm should evaluate. 
LocalStack supports metric alarms with simple statics. This typically includes:
* **metric-name**
* **namespace**
* **threshold** - a double value
* **comparison-operator** - e.g. _GreaterThanThreshold_, _LessThanThreshold_, _GreaterThanOrEqualToThreshold_, _LessThanOrEqualToThreshold_
* **statistic** - e.g. _SampleCount_, _Average_, _Sum_, _Minimum_, _Maximum_
* **period** - the interval in seconds used to calculate the metric
* **evaluation-periods** - defines the number of periods that will be evaluated to check the threshold
* **treat-missing-data** - defines how missing data points should be evaluated: _missing_, _ignore_, _notBreaching_, _breaching_

For a deeper understanding of alarm evaluation, we advice to consult the official [AWS docs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarm-evaluation).

Even though LocalStack's implementation of CloudWatch is not yet feature complete, you can already cover a range of use-cases. Next, we look into how we can make use of CloudWatch alarms with Lambdas.

## Setup a Lambda Function

> For the sake of simplicity, we will create a Lambda that will always fail. This will make it easier to demonstrate the alarms functionality. If you are interested in further learning about Lambdas, you can check our previous blog post on [Hot Swapping Python Lambda Functions using LocalStack](../2022-03-07-hot-swapping-python-lambda-functions-using-localstack), which also gives a solid introduction to Lambdas.

Reviewer 2: ```suggestion
For simplicity, we will create a Lambda that will always fail. This will make it easier to demonstrate the alarm's functionality. Suppose you are interested in further learning about Lambdas. In that case, you can check our previous blog post on [Hot Swapping Python Lambda Functions using LocalStack](../2022-03-07-hot-swapping-python-lambda-functions-using-localstack), which also gives a solid introduction to Lambdas.
```

Now, we just need to think about reasonable evaluation criteria - what is reasonable will depend a lot on your specific use case.

Let's say we want to get notified if the lambda fails at least once within the past minute.
Then the `evaluation-periods` would be 1, and `period` would be 60. 

For the `statistic` we use _Sum_ (failing invocation within the last minute will be summed up), and our `threshold` is 1, with the `comparison-operator` _GreaterThanOrEqualToThreshold_. 

In case the lambda was not invoke, we assume everything is fine. So we set `treat-missing-data` to _notBreaching_.

At this time we know how we want to evaluate the metrics. So the only thing left is configuring the action that should be executed, once the alarm changes its state to _ALARM_.

## Alarm actions
Actions are optional parameters that can be configured for alarms. There is one configuration for every state: `alarm-actions`, `ok-actions`, and `insufficient-data-actions`. The action will be executed once the alarm state changes. 

LocalStack supports SNS Topics as actions. 

> The topic must be created beforehand, and we need to know the topic's ARN for the metric-alarm configuration.

Reviewer 2: ```suggestion
Actions are optional parameters that we can configure for alarms. There is one configuration for every state: `alarm-actions`, `ok-actions`, and `insufficient-data-actions`. We will execute the action once the alarm state changes. 

LocalStack supports SNS Topics as actions. We must create the topic beforehand, and we need to know the topic's ARN for the metric-alarm configuration.
```

```

The email will contain plain text, and gives you information about the alarm that was triggered:
```sh
Content-Type: text/plain; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: SNS-Subscriber-Endpoint
From: LocalStack <admin@localstack.com>
To: stefanie@example.com

{"OldStateValue": "OK", "AlarmName": "my-lambda-alarm", "AlarmDescription": "", "AlarmConfigurationUpdatedTimestamp": "2022-05-04T13:58:57.297769000Z", "NewStateValue": "ALARM", "NewStateReason": "Threshold crossed", "StateChangeTime": "2022-05-04T13:59:20.438156000Z", "Region": "us-east-1", "AlarmArn": "arn:aws:cloudwatch:us-east-1:000000000000:alarm:my-lambda-alarm", "OKActions": "", "AlarmActions": ["arn:aws:sns:us-east-1:000000000000:my-topic-alarm"], "InsufficientDataActions": "", "Trigger": {"MetricName": "Errors", "Namespace": "AWS/Lambda", "Unit": "", "Period": 60, "EvaluationPeriods": 1, "ComparisonOperator": "GreaterThanOrEqualToThreshold", "Threshold": 1.0, "TreatMissingData": "notBreaching", "EvaluateLowSampleCountPercentile": "", "Dimensions": [{"value": "my-failing-lambda", "name": "FunctionName"}], "StatisticType": "Statistic", "Statistic": "SUM"}}
```

## Conclusion

> CloudWatch is an integral part of infrastructure management to monitor and react to operational metrics. Metric alarms are periodically evaluated and make you aware if thresholds of metrics are breached.
> 
> The configured actions trigger automatically once the alarm state changes and can therefore help to respond quick to any anomalies.
> 
> In our example we showed how to use email notifications, but this is just one way to do it. You could also use webhooks, SQS queues, or other SNS topic subscribers.

Reviewer 2: ```suggestion
CloudWatch is an integral part of infrastructure management to monitor and react to operational metrics. Metric alarms are periodically evaluated and make you aware if thresholds of metrics are breached.
		
The configured actions automatically trigger once the alarm state changes and can help respond quickly to any anomalies.
		
In our example, we showed how to use email notifications, but this is just one way. You could also use webhooks, SQS queues, or other SNS topic subscribers.
```
